src/Feldspar/Onnx/Operators.hs

```diff
@@ -284,20 +286,25 @@ takeDropShape i sh = P.splitAt j es
         j = if i P.< 0 then i + P.length es else i
 
 -- | Matrix multiplication of two dimensional temsors
-onnxGemm3 :: (RealFloat a, Numeric a, Pully vec, VecShape vec ~ DIM2)
-          => Attrs -> vec (Data a) -> vec (Data a) -> vec (Data a) -> DPull DIM2 a
-onnxGemm3 attrs vA vB vC = bcZipWith (+) (mmT vAT vBnT) $ toPull vC
-  where vA' = if alpha P.== 1.0 then toPull vA else map (* value alpha) $ toPull vA
+onnxGemm :: (RealFloat a, Numeric a, Pully vec, VecShape vec ~ DIM2, Pully vec2,
+             UnionShape DIM2 (VecShape vec2) ~ DIM2, Storable vec)
+         => Attrs -> vec (Data a) -> vec (Data a) -> vec2 (Data a) -> DPull DIM2 a
+onnxGemm attrs vA vB vC = bcZipWith (+) (mmT vAT vBnT) $ toPull vC
+  where vA' = if alpha P.== 1.0 then toPull $ store vA else map (* value alpha) $ toPull $ store vA
         vAT = if transA P.== 1 then transpose vA' else vA'
-        vB' = toPull vB
+        vB' = toPull $ store vB
         vBnT = if transA P.== transB then transpose vB' else vB' -- The mmT routine wants its second argument transposed
         alpha = P.realToFrac $ getAttr attrs aaFloat 1.0 "alpha"
         transA = getAttr attrs aaInt 0 "transA"
         transB = getAttr attrs aaInt 0 "transB"
 
 -- | Matrix multiplication that transposes its second argument
 mmT :: (Syntax a, Num a) => Pull DIM2 a -> Pull DIM2 a -> Pull DIM2 a
-mmT vecA vecBT = vvmap dim1 (\ rowA -> vvmap dim1 (sum . zipWith (*) rowA) vecBT) vecA
+-- mmT vecA vecBT = vvmap dim1 (\ rowA -> vvmap dim1 (sum . zipWith (*) rowA) vecBT) vecA
```

Comment:
Is this a workaround? Should the old code be removed?

Reply:
Yes, it should be removed. Done!